File: ui/web/src/lib/api/room.js
/**
 * Room 협업 세션 API 클라이언트 — 서버 /api/room/* 엔드포인트 1:1 래핑.
 */
import { BASE, authedFetch } from "./http.js";
import { withTokenQuery } from "./token.js";

/** 현재 룸 상태 조회. 룸이 없으면 null 반환. */
export async function roomState() {
	const res = await authedFetch(`${BASE}/api/room/state`);
	if (res.status === 404) return null;
	if (!res.ok) throw new Error(`room/state 실패: ${res.status}`);
	return res.json();
}

/** 룸 참여. { memberId, roomId, state } 반환. */
export async function roomJoin(name) {
	const res = await authedFetch(`${BASE}/api/room/join`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ name }),
	});
	if (!res.ok) throw new Error(`room/join 실패: ${res.status}`);
	return res.json();
}

/** 룸 퇴장. */
export async function roomLeave(memberId) {
	const res = await authedFetch(`${BASE}/api/room/leave`, {
		method: "POST",
		headers: _memberHeaders(memberId),
	});
	if (!res.ok) throw new Error(`room/leave 실패: ${res.status}`);
	return res.json();
}

/** 프레즌스 heartbeat (15초 간격). */
export async function roomHeartbeat(memberId) {
	const res = await authedFetch(`${BASE}/api/room/heartbeat`, {
		method: "POST",
		headers: _memberHeaders(memberId),
	});
	if (!res.ok) throw new Error(`room/heartbeat 실패: ${res.status}`);
	return res.json();
}

/** SSE 브로드캐스트 스트림. EventSource 반환. */
export function roomStream(memberId) {
	return new EventSource(withTokenQuery(`${BASE}/api/room/stream?member=${encodeURIComponent(memberId)}`));
}

/** AI 질문 → 전체 브로드캐스트. */
export async function roomAsk(memberId, question, company) {
	const body = { question };
	if (company) body.company = company;
	const res = await authedFetch(`${BASE}/api/room/ask`, {
		method: "POST",
		headers: { "Content-Type": "application/json", "x-room-member": memberId },
		body: JSON.stringify(body),
	});
	if (!res.ok) throw new Error(`room/ask 실패: ${res.status}`);
	return res.json();
}

/** 네비게이션 동기화. */
export async function roomNavigate(memberId, nav) {
	const res = await authedFetch(`${BASE}/api/room/navigate`, {
		method: "POST",
		headers: { "Content-Type": "application/json", "x-room-member": memberId },
		body: JSON.stringify(nav),
	});
	if (!res.ok) throw new Error(`room/navigate 실패: ${res.status}`);
	return res.json();
}

/** 채팅 메시지 전송. */
export async function roomChat(memberId, text) {
	const res = await authedFetch(`${BASE}/api/room/chat`, {
		method: "POST",
		headers: { "Content-Type": "application/json", "x-room-member": memberId },
		body: JSON.stringify({ text }),
	});
	if (!res.ok) throw new Error(`room/chat 실패: ${res.status}`);
	return res.json();
}

/** 이모지 반응. */
export async function roomReact(memberId, emoji, targetEvent) {
	const body = { emoji };
	if (targetEvent) body.targetEvent = targetEvent;
	const res = await authedFetch(`${BASE}/api/room/react`, {
		method: "POST",
		headers: { "Content-Type": "application/json", "x-room-member": memberId },
		body: JSON.stringify(body),
	});
	if (!res.ok) throw new Error(`room/react 실패: ${res.status}`);
	return res.json();
}

// ── 내부 헬퍼 ──

function _memberHeaders(memberId) {
	return { "Content-Type": "application/json", "x-room-member": memberId };
}

File: src/dartlab/server/api/room.py
"""Room 협업 세션 API — SSE fan-out + POST-back."""

from __future__ import annotations

import asyncio
import json
import logging
import time

from fastapi import APIRouter, HTTPException, Request
from sse_starlette.sse import EventSourceResponse

from ..models import (
    RoomAskRequest,
    RoomChatRequest,
    RoomJoinRequest,
    RoomNavigateRequest,
    RoomReactRequest,
)
from ..room import Room, RoomMember, room_manager

logger = logging.getLogger(__name__)

router = APIRouter()


# ---------------------------------------------------------------------------
# 헬퍼
# ---------------------------------------------------------------------------


def _getRoom() -> Room:
    room = room_manager.getRoom()
    if room is None:
        raise HTTPException(status_code=404, detail="협업 세션이 활성화되지 않았습니다.")
    return room


def _getMember(request: Request, room: Room) -> RoomMember:
    memberId = request.headers.get("x-room-member", "")
    member = room.getMember(memberId)
    if member is None:
        raise HTTPException(status_code=401, detail="룸에 참여하지 않은 사용자입니다.")
    return member


def _requireFull(member: RoomMember) -> None:
    if member.accessLevel != "full":
        raise HTTPException(status_code=403, detail="읽기 전용 토큰으로는 이 작업을 수행할 수 없습니다.")


# ---------------------------------------------------------------------------
# 엔드포인트
# ---------------------------------------------------------------------------


@router.post("/api/room/join")
async def roomJoin(req: RoomJoinRequest):
    """룸 참여 — member_id + 현재 상태 반환."""
    room = _getRoom()
    member = await room.join(req.name)
    if member is None:
        raise HTTPException(status_code=409, detail="룸 정원이 초과되었습니다.")

    state = room.getState()
    return {
        "memberId": member.memberId,
        "roomId": room.roomId,
        "members": state["members"],
        "navState": state["navState"],
        "chatHistory": state["chatHistory"],
    }


@router.post("/api/room/leave")
async def roomLeave(request: Request):
    """룸 퇴장."""
    room = _getRoom()
    member = _getMember(request, room)
    await room.leave(member.memberId)
    return {"status": "ok"}


@router.post("/api/room/heartbeat")
async def roomHeartbeat(request: Request):
    """프레즌스 유지."""
    room = _getRoom()
    memberId = request.headers.get("x-room-member", "")
    if not room.heartbeat(memberId):
        raise HTTPException(status_code=401, detail="룸에 참여하지 않은 사용자입니다.")
    return {"status": "ok", "members": len(room.members)}


@router.get("/api/room/state")
async def roomState():
    """현재 룸 상태."""
    room = _getRoom()
    return room.getState()


@router.get("/api/room/stream")
async def roomStream(request: Request):
    """SSE 스트림 — 브로드캐스트 수신."""
    room = _getRoom()
    memberId = request.query_params.get("member", "")
    member = room.getMember(memberId)
    if member is None:
        raise HTTPException(status_code=401, detail="룸에 참여하지 않은 사용자입니다.")

    async def _generate():
        try:
            while True:
                try:
                    msg = await asyncio.wait_for(member.queue.get(), timeout=30)
                    yield {
                        "event": msg["event"],
                        "data": json.dumps(msg["data"], ensure_ascii=False),
                    }
                except TimeoutError:
                    # keepalive — SSE comment
                    yield {"comment": "keepalive"}
        except asyncio.CancelledError:
            return

    return EventSourceResponse(_generate(), media_type="text/event-stream")


@router.post("/api/room/ask")
async def roomAsk(req: RoomAskRequest, request: Request):
    """질문 → 전체 브로드캐스트."""
    room = _getRoom()
    member = _getMember(request, room)
    _requireFull(member)

    if room._analyzing:
        raise HTTPException(status_code=409, detail="이미 분석이 진행 중입니다.")

    room._analyzing = True
    try:
        # 시작 브로드캐스트
        await room.broadcast(
            "ask_start",
            {
                "memberId": member.memberId,
                "name": member.name,
                "question": req.question,
                "company": req.company,
            },
        )

        # 스트리밍 인프라 — AI가 종목을 자율 판단
        from ..models import AskRequest
        from ..streaming import streamAsk

        ask_req = AskRequest(
            company=req.company,
            question=req.question,
            stream=True,
        )

        async for sse_event in streamAsk(ask_req):
            event_name = sse_event.get("event", "chunk")
            try:
                data = json.loads(sse_event.get("data", "{}"))
            except (json.JSONDecodeError, ValueError):
                data = {"raw": sse_event.get("data", "")}
            await room.broadcast(event_name, data)

            # 네비게이션 상태 업데이트 (meta 이벤트에서 종목 정보 추출)
            if event_name == "meta":
                room.nav_state.update({k: v for k, v in data.items() if k in ("stockCode", "corpName")})

    finally:
        room._analyzing = False

    return {"status": "ok"}


@router.post("/api/room/navigate")
async def roomNavigate(req: RoomNavigateRequest, request: Request):
    """네비게이션 동기화."""
    room = _getRoom()
    member = _getMember(request, room)
    _requireFull(member)

    nav_update = {k: v for k, v in req.model_dump().items() if v is not None}
    room.nav_state.update(nav_update)

    await room.broadcast(
        "navigate",
        {
            "memberId": member.memberId,
            "name": member.name,
            **nav_update,
        },
    )
    return {"status": "ok"}


@router.post("/api/room/chat")
async def roomChat(req: RoomChatRequest, request: Request):
    """채팅 메시지."""
    room = _getRoom()
    member = _getMember(request, room)

    msg = room.addChat(member.memberId, req.text)
    if msg is None:
        raise HTTPException(status_code=401, detail="룸에 참여하지 않은 사용자입니다.")

    await room.broadcast("chat", msg.toDict())
    return {"status": "ok"}


@router.post("/api/room/react")
async def roomReact(req: RoomReactRequest, request: Request):
    """이모지 반응."""
    room = _getRoom()
    member = _getMember(request, room)

    await room.broadcast(
        "react",
        {
            "memberId": member.memberId,
            "name": member.name,
            "emoji": req.emoji,
            "targetEvent": req.targetEvent,
            "timestamp": time.time(),
        },
    )
    return {"status": "ok"}
